File: src/main/java/link/pihda/billofmaterial/views/ItemListViewController.java
/*
 * Copyright (c) 2023.  Adib S. A.
 * Copyright (c) 2023.  PT Atadopos
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 */

package link.pihda.billofmaterial.views;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.fxml.FXMLLoader;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import link.pihda.billofmaterial.entity.Item;
import link.pihda.billofmaterial.entity.Procurement;
import link.pihda.billofmaterial.entity.Transaction;
import link.pihda.billofmaterial.enums.CurrencyCode;
import link.pihda.billofmaterial.service.ProcurementService;
import link.pihda.billofmaterial.service.TransactionService;
import link.pihda.billofmaterial.ui.RegexLimitingTextField;
import link.pihda.billofmaterial.util.ExcelGenerator;
import link.pihda.billofmaterial.util.InputValidator;

import java.text.NumberFormat;
import java.util.concurrent.atomic.AtomicInteger;

import static link.pihda.billofmaterial.util.GuiUtil.ChangeScene;
import static link.pihda.billofmaterial.util.TableUtil.createRowWithClickHandler;
import static link.pihda.billofmaterial.util.TableUtil.setDoubleCellFactory;

public final class ItemListViewController {

    // Services
    private final TransactionService transactionService = new TransactionService();
    private final ProcurementService procurementService = new ProcurementService();
    // State
    private final AtomicInteger lastSelectedIndex = new AtomicInteger(-1);
    // UI Components
    public TableView<Item> itemListTable;
    public TableColumn<Item, String> itemColumn;
    public TableColumn<Item, Double> quantityColumn;
    public TableColumn<Item, String> unitColumn;
    public TableColumn<Item, Double> priceColumn;
    public TableColumn<Item, Double> totalPriceColumn;
    public TextArea descriptionArea;
    public TextArea buyLinkArea;
    public Label descriptionLabel;
    public Label buyLinkLabel;
    public Label subtotalLabel;
    public Label taxesLabel;
    public Label shippingFeesLabel;
    public Label totalLabel;
    public ComboBox<CurrencyCode> currencyComboBox;
    public Button removeItemButton;
    public RegexLimitingTextField taxesInput;
    public RegexLimitingTextField shippingFeesInput;
    public Button backButton;
    public TextField transactionNameField;
    public Button editItemButton;
    private Transaction transaction;

    private Procurement procurement;

    // Method to get the FXML view
    public static FXMLLoader getView() {
        return new FXMLLoader(ItemListViewController.class.getResource("ItemListView.fxml"));
    }

    private void updateTransactionDetails() {
        if (transaction != null) {
            NumberFormat format = NumberFormat.getNumberInstance();
            format.setMinimumFractionDigits(2);
            format.setMaximumFractionDigits(2);

            // Load items and set table data
            ObservableList<Item> items = FXCollections.observableArrayList(transaction.getItems());
            itemListTable.setItems(items);

            // Calculating the subtotal by summing up total prices of all items.
            double subtotal = items.stream().mapToDouble(Item::getTotalPrice).sum();

            // Calculating the total cost of the transaction
            transaction.setTotalPrice(subtotal + transaction.getTaxes() + transaction.getShippingRate());
            transactionService.update(transaction);
            procurementService.update(procurement);

            // Display the details
            subtotalLabel.setText(format.format(subtotal));
            taxesLabel.setText(format.format(transaction.getTaxes()));
            shippingFeesLabel.setText(format.format(transaction.getShippingRate()));
            totalLabel.setText(format.format(transaction.getTotalPrice()));
        }
    }

    // Round to two decimal points


    // Method to initialize the BOM Table
    private void initializeTable() {
        taxesInput.setInputValidator(new InputValidator("^[0-9]*$", "Only numbers are allowed"));
        shippingFeesInput.setInputValidator(new InputValidator("^[0-9]*$", "Only numbers are allowed"));
        itemListTable.setColumnResizePolicy(TableView.CONSTRAINED_RESIZE_POLICY_FLEX_LAST_COLUMN);

        // Set cell value factories
        itemColumn.setCellValueFactory(new PropertyValueFactory<>("item"));
        quantityColumn.setCellValueFactory(new PropertyValueFactory<>("quantity"));
        unitColumn.setCellValueFactory(new PropertyValueFactory<>("unit"));
        priceColumn.setCellValueFactory(new PropertyValueFactory<>("price"));
        totalPriceColumn.setCellValueFactory(new PropertyValueFactory<>("totalPrice"));

        setDoubleCellFactory(priceColumn);
        setDoubleCellFactory(totalPriceColumn);

        // Set table selection listener
        itemListTable.getSelectionModel().selectedItemProperty().addListener(
                (obs, oldSelection, newSelection) -> handleTableSelection(newSelection)
        );

        // Set custom row factory to handle row clicks
        itemListTable.setRowFactory(tv -> createRowWithClickHandler(itemListTable, lastSelectedIndex));

        // Bind button disable property to table selection
        removeItemButton.disableProperty().bind(itemListTable.getSelectionModel().selectedItemProperty().isNull());
        editItemButton.disableProperty().bind(itemListTable.getSelectionModel().selectedItemProperty().isNull());

    }

    // Method to initialize currency ComboBox
    private void initializeCurrencyComboBox() {
        currencyComboBox.setItems(FXCollections.observableArrayList(CurrencyCode.values()));
    }

    // Method to initialize the information area
    private void initializeInfoArea() {
        collapseInfoArea(true);
    }


    // Method to handle table selection
    private void handleTableSelection(Item newSelection) {
        if (newSelection != null) {
            descriptionArea.setText(newSelection.getDescription());
            buyLinkArea.setText(newSelection.getBuyLink());
            collapseInfoArea(false);
        } else {
            collapseInfoArea(true);
        }
    }

    // Method to toggle the visibility of the information area
    private void collapseInfoArea(Boolean collapse) {
        boolean visible = !collapse;
        descriptionLabel.setVisible(visible);
        descriptionLabel.setManaged(visible);
        buyLinkLabel.setVisible(visible);
        buyLinkLabel.setManaged(visible);
        descriptionArea.setVisible(visible);
        descriptionArea.setManaged(visible);
        buyLinkArea.setVisible(visible);
        buyLinkArea.setManaged(visible);
    }

    // Method to add an item
    public void handleAddItem(ActionEvent actionEvent) {
        ItemFormController itemFormController = ChangeScene(ItemFormController.getView(), actionEvent);
        itemFormController.init(procurement, transaction);
    }

    // Method to remove an item
    public void handleRemoveItem(ActionEvent actionEvent) {
        Item selectedItem = itemListTable.getSelectionModel().getSelectedItem();
        if (selectedItem != null) {
            transaction.getItems().remove(selectedItem);
            transactionService.update(transaction);
            itemListTable.getItems().remove(selectedItem);
        } else {
            System.out.println("No item selected to remove.");
        }
    }

    public void updateTaxes(ActionEvent actionEvent) {
        transactionService.update(transaction.setTaxes(Double.parseDouble(taxesInput.getText())));
        updateTransactionDetails();
    }

    public void updateShipping(ActionEvent actionEvent) {
        transactionService.update(transaction.setShippingRate(Double.parseDouble(shippingFeesInput.getText())));
        updateTransactionDetails();
    }

    public void handleBack(ActionEvent actionEvent) {
        TransactionListViewController transactionListViewController = ChangeScene(TransactionListViewController.getView(), actionEvent);
        transactionListViewController.init(procurement);
    }

    public void updateCurrency(ActionEvent actionEvent) {
        transactionService.update(transaction.setCurrencyCode(currencyComboBox.getValue()));
    }

    public void handleChangeTransactionName(ActionEvent actionEvent) {
        String newTransactionName = transactionNameField.getText();
        if (newTransactionName != null && !newTransactionName.trim().isEmpty()) {
            transaction.setTransactionName(newTransactionName);
            transactionService.update(transaction);
        } else {
            transactionNameField.setText(transaction.getTransactionName());
        }
    }

    public void handleEditItem(ActionEvent actionEvent) {
        Item selectedItem = itemListTable.getSelectionModel().getSelectedItem();
        if (selectedItem != null) {
            ItemFormController itemFormController = ChangeScene(ItemFormController.getView(), actionEvent);
            itemFormController.init(procurement, transaction);
            itemFormController.setItem(selectedItem);
        } else {
            System.out.println("No item selected to update.");
        }
    }

    public void handleExportExcel(ActionEvent actionEvent) {
        ExcelGenerator.generateExcelFromItems(transaction.getItems());
    }

    public void init(Procurement procurement, Transaction transaction) {
        this.procurement = procurement;
        this.transaction = transaction;

        currencyComboBox.getSelectionModel().select(transaction.getCurrencyCode());
        transactionNameField.setText(transaction.getTransactionName());

        initializeTable();
        initializeCurrencyComboBox();
        initializeInfoArea();
        updateTransactionDetails();
    }
}

File: src/main/kotlin/link/pihda/billofmaterial/views/ItemListViewController.kt
/*
 * Copyright (c) 2023.  Adib S. A.
 * Copyright (c) 2023.  PT Atadopos
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 */
package link.pihda.billofmaterial.views

import javafx.collections.FXCollections
import javafx.event.ActionEvent
import javafx.fxml.FXMLLoader
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.control.cell.PropertyValueFactory
import javafx.util.Callback
import link.pihda.billofmaterial.entity.Item
import link.pihda.billofmaterial.entity.Procurement
import link.pihda.billofmaterial.entity.Transaction
import link.pihda.billofmaterial.enums.CurrencyCode
import link.pihda.billofmaterial.service.ProcurementService
import link.pihda.billofmaterial.service.TransactionService
import link.pihda.billofmaterial.ui.RegexLimitingTextField
import link.pihda.billofmaterial.util.ExcelGenerator
import link.pihda.billofmaterial.util.GuiUtil.changeScene
import link.pihda.billofmaterial.util.InputValidator
import link.pihda.billofmaterial.util.TableUtil.createRowWithClickHandler
import link.pihda.billofmaterial.util.TableUtil.setDoubleCellFactory
import java.text.NumberFormat
import java.util.concurrent.atomic.AtomicInteger

class ItemListViewController {
    // Services
    private val transactionService = TransactionService()
    private val procurementService = ProcurementService()

    // State
    private val lastSelectedIndex = AtomicInteger(-1)

    // UI Components
    lateinit var itemListTable: TableView<Item>
    lateinit var itemColumn: TableColumn<Item, String>
    lateinit var quantityColumn: TableColumn<Item, Double>
    lateinit var unitColumn: TableColumn<Item, String>
    lateinit var priceColumn: TableColumn<Item, Double>
    lateinit var totalPriceColumn: TableColumn<Item, Double>
    lateinit var descriptionArea: TextArea
    lateinit var buyLinkArea: TextArea
    lateinit var descriptionLabel: Label
    lateinit var buyLinkLabel: Label
    lateinit var subtotalLabel: Label
    lateinit var taxesLabel: Label
    lateinit var shippingFeesLabel: Label
    lateinit var totalLabel: Label
    lateinit var currencyComboBox: ComboBox<CurrencyCode>
    lateinit var removeItemButton: Button
    lateinit var taxesInput: RegexLimitingTextField
    lateinit var shippingFeesInput: RegexLimitingTextField
    lateinit var backButton: Button
    lateinit var transactionNameField: TextField
    lateinit var editItemButton: Button
    private lateinit var transaction: Transaction
    private lateinit var procurement: Procurement

    // Method to get the FXML view
    companion object {
        fun getView() = FXMLLoader(ItemListViewController::class.java.getResource("ItemListView.fxml"))
    }

    private fun updateTransactionDetails() {
        transaction.let {
            val format = NumberFormat.getNumberInstance().apply {
                minimumFractionDigits = 2
                maximumFractionDigits = 2
                maximumIntegerDigits = Int.MAX_VALUE
            }

            // Load items and set table data
            val items = FXCollections.observableArrayList(it.items)
            itemListTable.items = items

            // Calculating the subtotal by summing up total prices of all items.
            val subtotal = items.map(Item::totalPrice).sum()

            // Calculating the total cost of the transaction
            it.totalPrice = subtotal + it.taxes + it.shippingRate
            transactionService.update(it)
            procurementService.update(procurement)

            // Display the details
            subtotalLabel.text = format.format(subtotal)
            taxesLabel.text = format.format(it.taxes)
            shippingFeesLabel.text = format.format(it.shippingRate)
            totalLabel.text = format.format(it.totalPrice)
        }
    }

    // Method to initialize the BOM Table
    private fun initializeTable() {
        taxesInput.inputValidator = InputValidator("^[0-9]*$", "Only numbers are allowed")
        shippingFeesInput.inputValidator = InputValidator("^[0-9]*$", "Only numbers are allowed")
        itemListTable.columnResizePolicy = TableView.CONSTRAINED_RESIZE_POLICY_FLEX_LAST_COLUMN

        // Set cell value factories
        itemColumn.cellValueFactory = PropertyValueFactory("item")
        quantityColumn.cellValueFactory = PropertyValueFactory("quantity")
        unitColumn.cellValueFactory = PropertyValueFactory("unit")
        priceColumn.cellValueFactory = PropertyValueFactory("price")
        totalPriceColumn.cellValueFactory = PropertyValueFactory("totalPrice")

        setDoubleCellFactory(priceColumn)
        setDoubleCellFactory(totalPriceColumn)

        // Set table selection listener
        itemListTable.selectionModel.selectedItemProperty().addListener { _, _, newSelection ->
            handleTableSelection(newSelection)
        }

        // Set custom row factory to handle row clicks

        // Set custom row factory to handle row clicks
        itemListTable.rowFactory = Callback {
            createRowWithClickHandler(
                itemListTable,
                lastSelectedIndex, ::editItem
            )
        }
        // Bind button disable property to table selection
        removeItemButton.disableProperty().bind(itemListTable.selectionModel.selectedItemProperty().isNull)
        editItemButton.disableProperty().bind(itemListTable.selectionModel.selectedItemProperty().isNull)
    }

    // Method to initialize currency ComboBox
    private fun initializeCurrencyComboBox() {
        currencyComboBox.items = FXCollections.observableArrayList(*CurrencyCode.entries.toTypedArray())
    }

    // Method to initialize the information area
    private fun initializeInfoArea() {
        collapseInfoArea(true)
    }

    // Method to handle table selection
    private fun handleTableSelection(newSelection: Item?) {
        if (newSelection != null) {
            descriptionArea.text = newSelection.description
            buyLinkArea.text = newSelection.buyLink
            collapseInfoArea(false)
        } else {
            collapseInfoArea(true)
        }
    }

    // Method to toggle the visibility of the information area
    private fun collapseInfoArea(collapse: Boolean) {
        val visible = !collapse
        descriptionLabel.isVisible = visible
        descriptionLabel.isManaged = visible
        buyLinkLabel.isVisible = visible
        buyLinkLabel.isManaged = visible
        descriptionArea.isVisible = visible
        descriptionArea.isManaged = visible
        buyLinkArea.isVisible = visible
        buyLinkArea.isManaged = visible
    }

    // Method to add an item
    fun handleAddItem(actionEvent: ActionEvent) {
        val itemFormController = changeScene(ItemFormController.getView(), actionEvent) as ItemFormController
        itemFormController.init(procurement, transaction)
    }

    // Method to remove an item
    fun handleRemoveItem(actionEvent: ActionEvent) {
        val selectedItem = itemListTable.selectionModel.selectedItem
        if (selectedItem != null) {
            transaction.items.remove(selectedItem)
            itemListTable.items.remove(selectedItem)
            transactionService.update(transaction)
            procurementService.update(procurement)
        } else {
            println("No item selected to remove.")
        }
    }

    fun updateTaxes(actionEvent: ActionEvent) {
        transactionService.update(transaction.apply { taxes = taxesInput.text.toDouble() })
        updateTransactionDetails()
    }

    fun updateShipping(actionEvent: ActionEvent) {
        transactionService.update(transaction.apply { shippingRate = shippingFeesInput.text.toDouble() })
        updateTransactionDetails()
    }

    fun handleBack(actionEvent: ActionEvent) {
        val transactionListViewController =
            changeScene(TransactionListViewController.getView(), actionEvent) as TransactionListViewController
        transactionListViewController.init(procurement)
    }

    fun updateCurrency(actionEvent: ActionEvent) {
        transactionService.update(transaction.apply { currencyCode = currencyComboBox.value })
    }

    fun handleChangeTransactionName(actionEvent: ActionEvent) {
        val newTransactionName = transactionNameField.text
        if (!newTransactionName.isNullOrBlank()) {
            transaction.transactionName = newTransactionName
            transactionService.update(transaction)
        } else {
            transactionNameField.text = transaction.transactionName
        }
    }

    private fun editItem(item: Item, scene: Scene) {
        val itemFormController = changeScene(ItemFormController.getView(), scene) as ItemFormController
        itemFormController.init(procurement, transaction)
        itemFormController.setItem(item)
    }

    fun handleEditItem(actionEvent: ActionEvent) {
        val selectedItem = itemListTable.selectionModel.selectedItem
        if (selectedItem != null) {
            val itemFormController = changeScene(ItemFormController.getView(), actionEvent) as ItemFormController
            itemFormController.init(procurement, transaction)
            itemFormController.setItem(selectedItem)
        } else {
            println("No item selected to update.")
        }
    }

    fun handleExportExcel(actionEvent: ActionEvent) {
        ExcelGenerator.generateExcelFromTransaction(transaction)
    }

    fun init(procurement: Procurement, transaction: Transaction) {
        this.procurement = procurement
        this.transaction = transaction

        currencyComboBox.selectionModel.select(transaction.currencyCode)
        transactionNameField.text = transaction.transactionName

        initializeTable()
        initializeCurrencyComboBox()
        initializeInfoArea()
        updateTransactionDetails()
    }
}